[sqlalchemy] Ensure TrinoDialect.get_table_names only returns real tables

john-bodley opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #python-client in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description

Per the [SQLAlchemy documentation](https://docs.sqlalchemy.org/en/14/core/reflection.html#sqlalchemy.engine.reflection.Inspector.get_table_names) the `get_table_names` method is intended to return the real table names, i.e., without views. The PR updates the `get_table_names` to only return real table names whereas previously it returned real tables and views.

Note there were no prior integration tests for the `TrinoDialect.get_view_names` method so I added some to cover this logic and threw in a few extra for free which should cover all the various scenarios.

<!-- Provide a user-friendly explanation, keep it brief if it isn't user-visible. -->
## Non-technical explanation

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

> The `TrinoDialect.get_table_names` function logic has changed to adhere to the SQLAlchemy specification. Now only real table names are returned as opposed to both real table and view names.

```markdown
* Return only tables in SQLAlchemy  `get_table_names`. Previously, it contained views. ({issue}`266`)
```

hashhar commented on trino/sqlalchemy/dialect.py:
```diff
@@ -216,8 +217,9 @@ def get_view_names(self, connection: Connection, schema: str = None, **kw) -> Li
-            FROM "information_schema"."views"
+            FROM "information_schema"."tables"
             WHERE "table_schema" = :schema
+              AND "table_type" = 'VIEW'
```
Please add a test covering this (probably in `test_sqlalchemy_integration`).

You can create a VIEW in the memory connector and verify that it's not present in output of `get_table_names` and is present in `get_view_names`.

john-bodley replied:
@hashhar are you aware that there's no current unit/integration tests for either the `get_table_names` and `get_view_names` methods?

hashhar replied:
yes I do (unfortunately). SQLAlchemy integration made it in because it was useful but there hadn't been much time invested into it (until very recently).

hashhar reviewed (approved):
Thank you.

hashhar commented:
Also commit message to say:

```
Ensure SQLAlchemy get_table_names only returns real tables

Per the [SQLAlchemy
documentation](https://docs.sqlalchemy.org/en/14/core/reflection.html#sqlalchemy.engine.reflection.Inspector.get_table_names)
the get_table_names method is intended to return the real table names,
i.e., exclude views.

Furthermore querying the INFORMATION_SCHEMAS.VIEWS tables seems to be
sub-performant for determining which entities are views—I speculate this
is because all the view definitions need to be extracted. A more
performant approach is to use INFORMATION_SCHEMAS.TABLES with the
appropriate type filter.
```

john-bodley commented on tests/integration/test_sqlalchemy_integration.py:
```diff
@@ -20,8 +23,16 @@
-    engine = sqla.create_engine(f"trino://test@{host}:{port}/{request.param}",
-                                connect_args={"source": "test", "max_attempts": 1})
+
+    engine = sqla.create_engine(
```
I needed to explicitly set the schema as part of the engine otherwise Trino barfed when trying to create the views with the `Schema must be specified when session schema is not set` error.

john-bodley commented:
@hashhar I've addressed your comments. I addition to adding an integration test for this specific change I added additional tests which should cover the gamut of conditions for the `TrinoDialect.get_table_names` method.

mdesmet commented on tests/integration/test_sqlalchemy_integration.py:
```diff
@@ -368,3 +379,45 @@ def test_get_table_comment(trino_connection):
+        metadata.drop_all(engine)
+
+
+@patch('trino.sqlalchemy.dialect.TrinoDialect._get_default_schema_name')
```
Is this patch really necessary? We only set the catalog in the connection string, so I would expect the schema to be already `None`.

john-bodley replied:
@mdesmet per [here](https://github.com/trinodb/trino-python-client/blob/master/trino/sqlalchemy/dialect.py#L195) if the schema is `None` it tries to fallback to the default schema. This patch ensures that on line #195 `schema` is `None` which means the exception will be invoked.

john-bodley commented:
@mdesmet thanks for the review. I've addressed your comments.

hashhar reviewed (commented):
Sorry about post-merge review.

It's a future-looking comment - useful to address but not required today.
